cs169/snapcon
# 187153142 Implement payment in multiple currencies

### [Pivotal Tracker Link][tracker]


[tracker]: https://www.pivotaltracker.com/story/show/187197048

[tracker]: https://www.pivotaltracker.com/story/show/187153142

## What this PR does:


Unfortunately quite a large PR that completes the payment process. Ideally I could've split the db migrations and frontend items into separate PRs, but the migration requires changing the parameters for a lot of tests.

1. Update ticket_purchase and payment records to store a required currency field
2. Where amounts are calculated, it is updated to store the converted local currency, using convert_currency. I avoid changing existing logic for calculating total price supplied from Ticket model as much as possible.
3. Update payments page to display the correct converted amount and symbol
4. Update payments page to pass the correct currency when accessing stripe's API

### Include screenshots, videos, etc.

<img width="793" alt="image" src="https://github.com/cs169/snapcon/assets/45834630/24b0ac6c-6298-430a-9fda-1f78ca9f923b">


#### Who authored this PR?

Just me


### How should this PR be tested?

* Is there a deploy we can view?
* What do the specs/features test?
Added 1 test in features/ticket_purchases_spec.rb that does the complete payment flow from selecting a currency, proceeding to payment, and then completing it

Added several tests in payment and ticket_purchase models that validate the existence of currency.

* Are there edge cases to watch out for?

It's kind of difficult to know how I can improving testing. I think for now the most important part is ensuring that payments are stored correct. The features test case ensures that the ticket_purchase records the correct amount and currency, but it may need more comprehensive testing. Also, may need testing for currency conversion functionality in the event that an existing currency was used to complete a transaction, but that currency is later removed. Will need to check for interaction with the db

#### Are there any complications to deploying this?



Will need a migration

### Checklist:
